### DISCUSSION TITLE
Think variant general discussion

### DISCUSSION THREAD

**Discussion by zhen-zen** [zhen-zen/YogaSMC#87] [General]
Most model independent issues, including key event, will be linked here.

Reference: 

https://github.com/torvalds/linux/blob/master/drivers/platform/x86/thinkpad_acpi.c
https://github.com/torvalds/linux/blob/master/Documentation/admin-guide/laptops/thinkpad-acpi.rst

https://github.com/openbsd/src/blob/master/sys/dev/acpi/acpithinkpad.c
https://github.com/openbsd/src/blob/master/share/man/man4/acpithinkpad.4

https://linrunner.de/tlp/usage/tlp.html

**Comment by junaedahmed** [comment]
@zhen-zen , Could this be possible to set battery threshold for each battery separately as it is done in Think Vantage app on windows.
I don't know for sure but, I think setting threshold at a combined battery in dual battery setup is not working. At least in my case it is not working. I am currently using @benbender [BATX](https://github.com/benbender/x1c6-hackintosh/blob/experimental/EFI/OC/dsl/SSDT-BATX.dsl) SSDT for combined battery.

**Reply by zhen-zen** [reply]
The combined battery is a virtual one and the threshold setting is read from other methods. Currently separate control is available via `ioio`. You can check the commands at https://github.com/zhen-zen/YogaSMC/projects/2

**Reply by junaedahmed** [reply]
Hi @zhen-zen ,
I did check with ioio and it works by that route. but doesn't work with Prefpane. So I reported here to make it discrete.

**Reply by zhen-zen** [reply]
So that's just user interface improvements. How's that look like in vantage?

**Reply by junaedahmed** [reply]
In Vantage there is two seperate threshold setting field for each battery attached. So, I though that might be the problem why setting only one threshold not working on my setup. But setting this way works as intended.
```
ioio -s ThinkVPC Battery 0
ioio -s ThinkVPC setCMstart 50
ioio -s ThinkVPC setCMstop 60

ioio -s ThinkVPC Battery 1
ioio -s ThinkVPC setCMstart 50
ioio -s ThinkVPC setCMstop 60
```

**Reply by zhen-zen** [reply]
Thank you for the reply. I just deleted your screenshot since it contains serial numbers. I'm going to duplicate the start-stop box in prefpane. And what's the reading in ioreg for `BAT_ANY` (0) `BAT_PRIMARY` (1) and `BAT_SECONDARY` (2)? I suppose the threshold should match the primary and secondary one. `50 - 0x32` and `60 - 0x3C` for last 2 digits.

**Reply by junaedahmed** [reply]
Sorry for late reply. I think this is resolved with latest release. will test more and let you know.

**Comment by Ngdaz** [comment]
I can't set the battery threshold, it reset to 0 and 100 after turn off SMCpane, and its doesn't work with command line: "ioio -s ThinkVPC setCMstart 75", did i config it wrong
![image](https://user-images.githubusercontent.com/47729815/117520334-e7fbda00-afd1-11eb-92db-a4eb71776262.png)

**Reply by zhen-zen** [reply]
Did battery threshold work in Linux and Windows?

**Reply by Ngdaz** [reply]
yes it works on linux and windown

**Reply by Ngdaz** [reply]
I am using t490

**Reply by Ngdaz** [reply]
and kext for my laptop is from here: 
https://github.com/yusifsalam/t490-macos

**Reply by Ngdaz** [reply]
Do I need to turn off anything in here ? 
![image](https://user-images.githubusercontent.com/47729815/117536157-aa27a180-b023-11eb-8774-64495b837956.png)

**Reply by zhen-zen** [reply]
Please attach a copy of ioreg. For `ioio` installation, see https://github.com/RehabMan/OS-X-ioio

**Reply by Ngdaz** [reply]
what "copy of ioreg" mean, what is ioreg, sorry I am a noob, could you guide me on how to do it

**Reply by djsnakezinho** [reply]
Hi
In my X1 Carbon 7th I can't change battery thresholds either. Prefane can read them correctly but I can only change them in Windows (via Lenovo Vantage app). I also tried setting them with ioio and even though no error is reported nothing changes. Prefane reports ECAccess as RW and I can change all other settings I've tried (LEDs, keyboard backlight, fan speed). Could I be missing something?
Thanks

**Reply by djsnakezinho** [reply]
Hi again
Changing thresholds with ioio does work after all! I was trying to disable conservation mode by setting CMstop to 100 (that's what YogaSMC shows in PrefPane), but it turns out you have to set it to 0. I still can't make these changes in the GUI, but I've created a couple of Automator apps to enable and disable conservation mode by running ioio commands and all is good!

**Reply by Mifaxav** [reply]
Im using a T480 but still cant change the settings of my battery using gui. Havent tried the debug kext yet... Could you send the ioio commands to help me build my own gui version?

**Reply by djsnakezinho** [reply]
Hi

This is what I do to disable conservation mode:
ioio -s ThinkVPC Battery 0
ioio -s ThinkVPC setCMstart 0
ioio -s ThinkVPC setCMstop 0

And to enable (start charging at 55% and stop at 60%):
ioio -s ThinkVPC Battery 0
ioio -s ThinkVPC setCMstart 55
ioio -s ThinkVPC setCMstop 60
